File: src/java/foundgine-core/src/main/java/com/foundgine/core/abstractions/AuthorizationDecision.java
package com.foundgine.core.abstractions;

import java.util.Objects;

/**
 * Port of {@code Foundgine.Core.Abstractions.AuthorizationDecision}.
 *
 * <p>Provider-independent authorization result. Conditional access carries a
 * predicate that must remain part of execution semantics and be evaluated
 * at the provider boundary with the current execution context.
 */
public record AuthorizationDecision(AuthorizationAccess access, AuthorizationPredicate predicate) {

    public AuthorizationDecision(AuthorizationAccess access) {
        this(access, null);
    }

    public static final AuthorizationDecision DENIED = new AuthorizationDecision(AuthorizationAccess.DENIED);
    public static final AuthorizationDecision ALLOWED = new AuthorizationDecision(AuthorizationAccess.ALLOWED);

    public static AuthorizationDecision conditional(AuthorizationPredicate predicate) {
        Objects.requireNonNull(predicate, "predicate");
        return new AuthorizationDecision(AuthorizationAccess.CONDITIONAL, predicate);
    }

    public boolean isAllowed() {
        return access == AuthorizationAccess.ALLOWED || access == AuthorizationAccess.CONDITIONAL;
    }

    public static AuthorizationDecision combine(AuthorizationDecision left, AuthorizationDecision right) {
        Objects.requireNonNull(left, "left");
        Objects.requireNonNull(right, "right");

        if (!left.isAllowed() || !right.isAllowed()) {
            return DENIED;
        }

        if (left.predicate() == null) {
            return right;
        }

        if (right.predicate() == null) {
            return left;
        }

        return conditional(AuthorizationPredicate.and(left.predicate(), right.predicate()));
    }
}

File: src/csharp/Foundgine.Core/Interfaces/AuthorizationDecision.cs
namespace Foundgine.Core.Abstractions;

/// <summary>Operation for which a semantic authorization decision is requested.</summary>
public enum AuthorizationOperation : byte
{
    Read,
    Write
}

/// <summary>High-level authorization state exposed by the semantic capability model.</summary>
public enum AuthorizationAccess : byte
{
    Denied,
    Allowed,
    Conditional
}

/// <summary>
/// Provider-independent authorization result. Conditional access carries a
/// predicate that must remain part of execution semantics and be evaluated at
/// the provider boundary with the current execution context.
/// </summary>
public sealed record AuthorizationDecision(
    AuthorizationAccess Access,
    AuthorizationPredicate? Predicate = null)
{
    public static AuthorizationDecision Denied { get; } = new(AuthorizationAccess.Denied);
    public static AuthorizationDecision Allowed { get; } = new(AuthorizationAccess.Allowed);

    public static AuthorizationDecision Conditional(AuthorizationPredicate predicate)
    {
        ArgumentNullException.ThrowIfNull(predicate);
        return new AuthorizationDecision(AuthorizationAccess.Conditional, predicate);
    }

    public bool IsAllowed => Access is AuthorizationAccess.Allowed or AuthorizationAccess.Conditional;

    public static AuthorizationDecision Combine(AuthorizationDecision left, AuthorizationDecision right)
    {
        ArgumentNullException.ThrowIfNull(left);
        ArgumentNullException.ThrowIfNull(right);

        if (!left.IsAllowed || !right.IsAllowed)
            return Denied;

        if (left.Predicate is null)
            return right;

        if (right.Predicate is null)
            return left;

        return Conditional(AuthorizationPredicate.And(left.Predicate, right.Predicate));
    }
}